src/main.rs

```diff
@@ -36,28 +36,34 @@ struct Query {
 #[post("/", format = "application/json", data = "<q>")]
 fn index(q: Json<Query>) -> Json<Value> {
     SYNTAX_SET.with(|syntax_set| {
-	    // Determine theme to use.
-	    //
-	    // TODO(slimsag): We could let the query specify the theme file's actual
-	    // bytes? e.g. via `load_from_reader`.
-	    let theme = match THEME_SET.themes.get(&q.theme) {
-	        Some(v) => v,
-	        None => return Json(json!({"error": "invalid theme"})),
-	    };
-
-	    // Determine syntax definition by extension.
-	    let syntax_def = match syntax_set.find_syntax_by_extension(&q.extension) {
-	        Some(v) => v,
-	        None => return Json(json!({"error": "invalid extension"})),
-	    };
-
-	    // TODO(slimsag): return the theme's background color (and other info??) to caller?
-	    // https://github.com/trishume/syntect/blob/c8b47758a3872d478c7fc740782cd468b2c0a96b/examples/synhtml.rs#L24
-
-	    Json(json!({
+        // Determine theme to use.
+        //
+        // TODO(slimsag): We could let the query specify the theme file's actual
+        // bytes? e.g. via `load_from_reader`.
+        let theme = match THEME_SET.themes.get(&q.theme) {
+            Some(v) => v,
+            None => return Json(json!({"error": "invalid theme"})),
+        };
+
+        // Fall back: Determine syntax definition by first line.
+        let syntax_by_first_line_def = match syntax_set.find_syntax_by_first_line(&q.code) {
+            Some(v) => v,
+            None => return Json(json!({"error": "invalid syntax"})),
```

Comment:
If finding the syntax by the first line fails here, we're always going to return `invalid syntax` and never actually get to finding by extension below. That's not good. Also, this code runs even if `find_syntax_by_extension` succeeds.

A better solution is to put this code inside `None => { ... here ...}`